### PR TITLE
Added support for logging responses of streaming calls.

### PR DESCRIPTION
**Pull request by fiboknacky**
Once this is LGTMed by the reviewers, I'll add support for v3 to v5.

### CODE REVIEW

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php [hunk 1]
```diff
@@ -44,15 +45,44 @@ public function __construct(
...
+        }
+    }
+
+    /**
+     * Logs summary and the details of the given status and response.
+     *
+     * @param object|null $status the status to be logged
+     * @param object|null $response the response to be logged
+     */
+    private function logSummaryAndDetails(object $status, ?object $response = null)
```
Comment (PierrickVoulet) [inline src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:84]: Should not `$status` be of type `?object` instead of `object`? I am asking because you use `object|null` in the comment block.

File: tests/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCallTest.php [hunk 2]
```diff
@@ -27,39 +30,112 @@
...
+
+    /**
+     * @depends testResponses
+     */
+    public function testGetStatus($dependingData)
+    {
+        $this->performStatusAssertion($dependingData);
+    }
+
+    public function testResponsesWithEmptyResponses()
```
Comment (PierrickVoulet) [inline tests/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCallTest.php:75]: I think you should replace this test with additional test(s) in the `LogMessageFormatterTest` that are based on `ServerStreamingCall` mock (see what is currently done for [UnaryCall](https://github.com/googleads/google-ads-php/blob/master/tests/Google/Ads/GoogleAds/Lib/V6/LogMessageFormatterTest.php#L678)).

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:84]: It should be object. $status shouldn't be null. I've updated the PHPDoc.

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php [hunk 3]
```diff
@@ -44,15 +45,44 @@ public function __construct(
...
+        return $status;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function responses()
+    {
+        foreach ($this->innerCall->responses() as $response) {
+            $this->savedResponses[] = $response;
```
[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:73]: There is no way we can do this accurately until we get the status.
The reason is we're relying on it to determine the correct log level. 
We can base the decision on either end case (level of logDetail when status code is not 0 or level of logSummary when status code is 0).
But those can still be confusing for people.

The fastest thing that I could do would be checking if the logging is enabled at all.
WDYT?

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:84]: Sounds good. Thanks for suggestion.

[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:84]: SG. Thanks for suggestion.

File: tests/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCallTest.php (hunk 2 above)
Comment (fiboknacky) [reply tests/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCallTest.php:75]: I can add more test in the LogMessageFormatterTest, but this still needs to be here.
The reason is we need to test getStatus when the response is empty. Without this method to depend on, we'll need to copy most code from the testResponses() into getStatus(), which I want to avoid.

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (PierrickVoulet) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:73]: What about adding a new setting that specifies the maximum number of responses that will be logged. We could then have a default value set to something reasonable for performance like 10 or 100?

File: tests/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCallTest.php (hunk 2 above)
Comment (PierrickVoulet) [reply tests/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCallTest.php:75]: SG. I was suggesting new tests because I believe the [empty/non-empty array logic](https://github.com/googleads/google-ads-php/pull/474/files#diff-d4a23076f7ae8234187fabb56741ae0dd14ca146afcd1f55fd2776a60dfbd9e7R57) is not entirely covered as of now.

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:73]: I can add it, but from what I test, most of the time, we got only one large response though.
So, not sure setting a default value like you mention would be applicable to this case? 

Let's hear from Anash too. I actually wonder how other libs deal with this.

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsCallLogger.php [hunk 4]
```diff
@@ -65,6 +65,25 @@ public function __construct(
...
     }
 
+    /**
+     * Logs summary and the details of the given status, request data and response.
+     *
+     * @param object $status the status to be logged
+     * @param array $requestData the request data
+     * @param object|null $response the response to be logged
+     */
+    public function log(object $status, array $requestData, ?object $response = null)
```
Comment (PierrickVoulet) [inline src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsCallLogger.php:75]: I believe you are missing a `call` parameter to avoid the use of `$this`?

Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsCallLogger.php]: Good call.
Fixed and confirmed that it worked.

**Review by PierrickVoulet — COMMENTED** [review]
The implementation looks good to me but I think we should get a consensus about [the last opened comment](https://github.com/googleads/google-ads-php/pull/474/files#r532718697) before approval.

File: src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php (hunk 3 above)
[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php:73]: > The fastest thing that I could do would be checking if the logging is enabled at all.

Seems this is not possible, as we always create a default logger with log level of DEBUG.

Other options would be:

1) Refactor the GoogleAdsCallLogger class to have a public method that can tell if the logger is enabled for DEBUG (the level that we use for logging details() in the case when there is no error).
2) Have a setting like you mention to limit the number of responses that we log. Users can change this later.

I've just pinged Anash to hear his opinions.

[the remark this answers is not in the document]
Comment (fiboknacky) [reply src/Google/Ads/GoogleAds/Lib/V6/GoogleAdsLoggingServerStreamingCall.php]: BTW, I just added a condition to store these responses only when the logger is enabled for logging those responses.

**Review by AnashOommen — APPROVED** [review]

**Review by PierrickVoulet — APPROVED** [review]

**Comment by fiboknacky** [comment]
Pierrick, could you take a quick look again? I've just copied and changed the version number for V3-V5. Thanks.

**Review by PierrickVoulet — APPROVED** [review]
